hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/ListingBasedRollbackStrategy.java

```diff
@@ -57,20 +81,252 @@ public ListingBasedRollbackStrategy(HoodieTable table,
+              // taken in this scenario is a combination of (A.2) and (A.3)
+              // ---------------------------------------------------------------------------------------------------
+              // (B) The following cases are possible if !index.canIndexLogFiles and/or !index.isGlobal
+              // ---------------------------------------------------------------------------------------------------
+              // (B.1) Failed first commit - Inserts were written to base files and HoodieWriteStat has no entries.
+              // In this scenario, we delete all the base files written for the failed commit.
+              // (B.2) Failed recurring commits - Inserts were written to base files and updates to log files. In
+              // this scenario, perform (A.1) and for updates written to log files, write rollback blocks.
+              // (B.3) Rollback triggered for first commit - Same as (B.1)
+              // (B.4) Rollback triggered for recurring commits - Same as (B.2) plus we need to delete the log files
+              // as well if the base base file gets deleted.
+              HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(
+                  table.getMetaClient().getCommitTimeline().getInstantDetails(instantToRollback).get(),
+                  HoodieCommitMetadata.class);
+
+              // In case all data was inserts and the commit failed, delete the file belonging to that commit
+              // We do not know fileIds for inserts (first inserts are either log files or base files),
+              // delete all files for the corresponding failed commit, if present (same as COW)
+              hoodieRollbackRequests.add(
+                  deleteDataAndLogFilesRollbackRequest(instantToRollback, metaClient, partitionPath));
+
+              // append rollback blocks for updates and inserts as A.2 and B.2
+              if (commitMetadata.getPartitionToWriteStats().containsKey(partitionPath)) {
+                hoodieRollbackRequests.addAll(
+                    getRollbackRequestForAppend(partitionPath, instantToRollback, commitMetadata, table));
+              }
+              break;
+            default:
+              throw new HoodieRollbackException("Unknown listing type, during rollback of " + instantToRollback);
+          }
+        }
+        return hoodieRollbackRequests.stream();
+      }, numPartitions);
+    } catch (Exception e) {
       LOG.error("Generating rollback requests failed for " + instantToRollback.getTimestamp(), e);
       throw new HoodieRollbackException("Generating rollback requests failed for " + instantToRollback.getTimestamp(), e);
     }
   }
+
+  private String getBaseFileExtension(HoodieTableMetaClient metaClient) {
+    return metaClient.getTableConfig().getBaseFileFormat().getFileExtension();
+  }
+
+  @NotNull
+  private HoodieRollbackRequest deleteDataAndLogFilesRollbackRequest(HoodieInstant instantToRollback,
```

Comment:
We should inline those methods, given that we're going to be supporting both rollbacks of completed instance as well as pending ones: 

```
getHoodieRollbackRequest(partitionPath, getBaseAndLogFilesToBeDeleted(metaClient, instantToRollback, partitionPath))
```

will become (for completed instants)

```
getHoodieRollbackRequest(partitionPath, fetchFilesFromCommitMetadata(metaClient, instantToRollback, partitionPath))
```

for pending ones

```
getHoodieRollbackRequest(partitionPath, listFilesToBeDeleted(instantToRollback.getTimestamp(), basefileExtension, partitionPath, fs))
```